hilti/toolchain/src/compiler/optimizer/passes/flatten-blocks.cc

```diff
@@ -0,0 +1,158 @@
+                id = ID(id.str() + "_");
+
+            used.emplace(id, decl);
+
+            // No need to trigger ID change if there was no conflict.
+            if ( id == decl->id() )
+                continue;
+
+            recordChange(decl, util::fmt(R"(renaming declaration "%s" -> "%s")", decl->id(), id));
+            decl->setID(id);
+
+            for ( auto* name : uses ) {
+                recordChange(name, util::fmt(R"(renaming reference "%s" -> "%s")", name->id(), id));
+                name->setID(id);
+            }
+        }
+
+        // Fold contents of block into parent.
+        auto contents = parent->children();
+        recordChange(parent, "inlining child block");
+        parent->clearChildren();
+
+        // Variables declared in the block would have previously gone
+        // out of scope. Overwrite them to force any aliases to also
+        // see an update.
+        auto* cfg = state()->cfg(function->body());
+        if ( ! contents.empty() ) {
+            if ( auto* last = *contents.rbegin() ) {
+                const auto& successors = cfg->graph().neighborsDownstream(last->identity());
+                // A block should have at most one child, the statement following it.
+                assert(successors.size() <= 1);
+                if ( ! successors.empty() ) {
+                    const auto* scope_end = cfg->graph().getNode(successors.front());
+                    assert(scope_end);
+                    assert((*scope_end)->isA<cfg::End>());
+                    const auto& transfer = cfg->dataflow().at(*scope_end);
+                    for ( auto&& [a, xx] : transfer.kill ) {
+                        // Since loops and conditionals can have a block we check whether the block actually
+                        // contains the declaration so we do not lifecycle variables declared in e.g., the loop
+                        // control block.
+                        if ( auto* local = a->tryAs<declaration::LocalVariable>();
+                             local && cfg::contains(*block, *local) ) {
+                            recordChange(local,
+                                         "resetting block-local variable at the end of block since block will "
+                                         "be removed");
+                            block->addChild(context(), builder()->assign(builder()->id(local->id()),
+                                                                         builder()->default_(local->type()->type())));
+                        }
+                    }
+                }
+            }
+        }
+
+        // Copy the original contents in order, but inline the block.
+        for ( auto* c : contents ) {
+            if ( c == block ) {
+                recordChange(block, "inlining block");
+
+                auto children = block->children();
+                block->clearChildren();
+                parent->addChildren(context(), children);
```

Comment:
Loop over child statements and use `parent->add(stmt)`, or add a new helper  method `add(Statements stmts)` to `Block`.